### PR TITLE
RSDK-2489

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/RSDK-2489)

Fixed issues:
1. If GetPosition or GetPointCloudMap returns an error, both the 2d & 3d slam views stop polling, reset the polling interval dropdown & pins the error that stopped polling to the top of the 2d or 3d slam view (respectively)
2. Makes it so that every time the 2d or 3d views are expanded 

TODO:
- [x] Test on fake slam (by triggering errors on an interval for GetPosition or GetPointCloud requests)
- [x] Test on rover
- [x] Confirm errors on get position also stop map polling

Screenshot 2023-04-21 at 4 19 22 PM
Screenshot 2023-04-21 at 4 20 40 PM

https://user-images.githubusercontent.com/5927876/233727611-165a812f-81dc-4693-801d-362a9314de82.mp4

### CODE REVIEW

File: services/slam/fake/data_loader.go [hunk 1]
```diff
@@ -54,12 +55,19 @@ func fakeGetPointCloudMap(_ context.Context, datasetDir string, slamSvc *SLAM) (
...
 	}
 	chunk := make([]byte, chunkSizeBytes)
+	var chunkCount int
 	f := func() ([]byte, error) {
 		bytesRead, err := file.Read(chunk)
 		if err != nil {
 			defer utils.UncheckedErrorFunc(file.Close)
 			return nil, err
 		}
+		chunkCount++
```
Comment (nicksanford) [inline services/slam/fake/data_loader.go:65]: remove before merge

File: web/frontend/src/components/slam.vue [hunk 2]
```diff
@@ -54,12 +56,16 @@ const fetchSLAMMap = (name: string): Promise<Uint8Array> => {
 
     const getPointCloudMap: ResponseStream<slamApi.GetPointCloudMapResponse> =
       props.client.slamService.getPointCloudMap(req);
+    let dataCounter = 0;
```
Comment (nicksanford) [inline web/frontend/src/components/slam.vue:59]: remove before merge

File: web/frontend/src/components/slam.vue [hunk 3]
```diff
@@ -54,12 +56,16 @@ const fetchSLAMMap = (name: string): Promise<Uint8Array> => {
...
     getPointCloudMap.on('data', (res: { getPointCloudPcdChunk_asU8(): Uint8Array }) => {
+      dataCounter += 1;
+      console.log('received data', dataCounter);
       const chunk = res.getPointCloudPcdChunk_asU8();
       chunks.push(chunk);
     });
     getPointCloudMap.on('status', (status: { code: number, details: string, metadata: grpc.Metadata }) => {
       if (status.code !== 0) {
+        console.warn('status called with status', status);
```
Comment (nicksanford) [inline web/frontend/src/components/slam.vue:68]: remove before merge

File: web/frontend/src/components/slam.vue [hunk 4]
```diff
@@ -68,10 +74,18 @@ const fetchSLAMMap = (name: string): Promise<Uint8Array> => {
...
-    getPointCloudMap.on('end', (end?: { code: number }) => {
-      if (end === undefined || end.code !== 0) {
-        // the error will be logged in the 'status' callback
-        return;
+    getPointCloudMap.on('end', (end?: { code: number, details: string, metadata: grpc.Metadata }) => {
+      if (end === undefined) {
+        const error = { message: 'Stream ended without status code' };
+        reject(error);
+      } else if (end.code !== 0) {
+        console.warn('called with end', end);
```
Comment (nicksanford) [inline web/frontend/src/components/slam.vue:82]: remove before merge

File: web/frontend/src/components/slam.vue [hunk 5]
```diff
@@ -249,7 +280,7 @@ onUnmounted(() => {
     />
     <div class="border-border-1 h-auto border-x border-b p-2">
       <div class="container mx-auto">
-        <div class="pt-4">
+        <div class="flex-col pt-4"> <!--here is the end of the thing-->
```
Comment (EshaMaharishi) [inline web/frontend/src/components/slam.vue]: is "" for debugging or should it be committed?

File: web/frontend/src/components/slam.vue [hunk 6]
```diff
@@ -68,10 +74,18 @@ const fetchSLAMMap = (name: string): Promise<Uint8Array> => {
         reject(error);
       }
     });
-    getPointCloudMap.on('end', (end?: { code: number }) => {
-      if (end === undefined || end.code !== 0) {
-        // the error will be logged in the 'status' callback
-        return;
+    getPointCloudMap.on('end', (end?: { code: number, details: string, metadata: grpc.Metadata }) => {
```
Comment (EshaMaharishi) [inline web/frontend/src/components/slam.vue:77]: just for my knowledge, do you know what this `(end?: { ... })` syntax means? It looks like some form of [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining), but I don't quite get how it works here

File: services/slam/fake/data_loader.go [hunk 7]
```diff
@@ -47,19 +47,29 @@ const (
...
+		slamSvc.logger.Warnf("slamSvc.getCount() > %d, chunkCount %d", slamSvc.getCount(), chunkCount)
+		if c%3 == 0 && chunkCount >= 2 {
+			return nil, fmt.Errorf("simulated bad PCD chunk %d", c)
+		}
```
Comment (EshaMaharishi) [inline services/slam/fake/data_loader.go]: will this also be removed? if not, do we want a simulated error both here and above?

File: services/slam/fake/data_loader.go [hunk 8]
```diff
@@ -47,19 +47,29 @@ const (
 )
 
 func fakeGetPointCloudMap(_ context.Context, datasetDir string, slamSvc *SLAM) (func() ([]byte, error), error) {
+	c := slamSvc.getCount()
+	if c == 17 {
+		return nil, fmt.Errorf("simulated bad fakeGetPointCloudMap response %d", c)
+	}
```
Comment (EshaMaharishi) [inline services/slam/fake/data_loader.go]: do these simulated errors mean fake will always produce errors?

**Review by EshaMaharishi — COMMENTED** [review]
changes to the web UI look great! i mainly had some questions on the changes to fake

File: services/slam/fake/data_loader.go (hunk 8 above)
Comment (nicksanford) [reply services/slam/fake/data_loader.go]: Yes I've removed them.

File: services/slam/fake/data_loader.go (hunk 7 above)
Comment (nicksanford) [reply services/slam/fake/data_loader.go]: Yes I've removed them.

File: web/frontend/src/components/slam.vue (hunk 6 above)
Comment (nicksanford) [reply web/frontend/src/components/slam.vue:77]: It means that end can be `null` 
https://stackoverflow.com/questions/37632760/what-is-the-question-mark-for-in-a-typescript-parameter-name#:~:text=Not%20exactly.-,Question%20mark%20means%20%22optional%22.,%7C%20undefined%20%3D%20undefined%22..

File: web/frontend/src/components/slam.vue (hunk 5 above)
Comment (nicksanford) [reply web/frontend/src/components/slam.vue]: removed

File: web/frontend/src/components/slam.vue (hunk 6 above)
Comment (EshaMaharishi) [reply web/frontend/src/components/slam.vue:77]: nice, ty!

additionally spoke offline that we think based on talking with max on the sdk team that end would be null if the server didn't respond, but would be populated if the server responded with an error.

further, if the callback passed to getPointCloudMap.on calls neither reject nor resolve, then the logical thread of execution will hang at [this](https://github.com/viamrobotics/rdk/blob/6d097d9e2905c4ef9bedc9aa619aff3c4b38a39a/web/frontend/src/components/slam.vue#L146) line

**Pull request merged**